README

 

                             . syck .

                         [ version 0.70 ]





INSTALLATION

   ./bootstrap
   ./configure
   make
   make check
   sudo make install

If the unit tests don't pass, notify me immediately.  This distribution
is tested on FreeBSD and Linux.  I don't release it unless the tests
pass on those machines.  If tests aren't passing, then that's a problem.

ABOUT

Syck is the Scripters' YAML Cobble-Yourself-a-Parser Kit.  I don't
much care if the acronym works, as long as the library does!

The whole point of Syck is to make parsing and emitting YAML very
simple for scripting languages through C bindings.  It doesn't strive
to be a pull parser or very extendible.  It just is concerned with
loading a YAML document into a C structure which can be easily 
translated into a scripting language's internal native data type.

RUBY INSTALLATION

You don't need to `make install', but please configure and make libsyck
as outlined above.

   cd ext/ruby
   ruby install.rb config
   ruby install.rb setup
   sudo ruby install.rb install

Syck works best with Ruby.  Ruby's symbol table is leveraged, as well
as Ruby's VALUE system.  (You can read more about that below.)

Syck is now included with Ruby (beginning with Ruby 1.8.0.)  Please
voice your support for Syck/YAML in Ruby distributions on the various
platforms.

PYTHON INSTALLATION

You'll need to `make install' as described above.

   cd ext/python/
   python setup.py build
   sudo python setup.py install

PHP INSTALLATION

You'll need to `make install' as described above.

   ln -s lib include    # or cp -r lib include
   cd ext/php/
   phpize
   ./configure --with-syck=../..
   make
   sudo make install

HOW SYCK IS SO GREAT

For example, in Ruby everything evaluates to a VALUE.  I merely
supply a handler to Syck that will take a SyckNode and transform
it into a Ruby VALUE.

A simple Ruby YAML::load could be built like so:

  static VALUE
  YAML_load( VALUE str )
  {
    SyckParser* parser;
    parser = syck_new_parser();
    syck_parser_handler( parser, YAML_handler );
    return syck_parse( parser, str );
  }

  static VALUE
  YAML_handler( SyckNode* node )
  {
    switch( node->kind )
    {
      case SYCK_MAP:
        VALUE key;
        VALUE h = rb_hash_new();
        for ( key = node->content[0]; key != null; key++ )
        {
          rb_hash_set( h, key, key++ );
        }
        return h;
      break;
    }
  }

For most C developers, it should be a no-brainer to bring
basic YAML serialization to PHP, Tcl, Cocoa, etc.

Instructions for using Syck's API are available in the
README.EXT in this very same directory.


CHANGELOG
--- %YAML:1.0
- version: 0.70
  date: 2009-09-30
  changes:
    - 64-bit compatibility

- version: 0.60
  date: 2005-09-19
  changes:
    lib/emitter.c: get headless documents working.  better quoting of various
      indicators at end or alone in scalars.
    lib/token.re: forgive tabs in plain scalars.
    ext/syck/rubyext.c: headless document always.

- version: 0.56
  date: 2005-05-16
  changes:
    ext/ruby/ext/syck/rubyext.c: 
    - const_find, now locating class constants correctly.
    - YAML::Object class for loaded objects which have no corresponding class.
    - No anchors on simple strings.
    - Outputing of domain and private types and anchors properly.
    - Memory leak in mktime.

    lib/emitter.c: scale back seq-in-map on a typed seq.  prevents the shift/reduce
      problem with the parser.

- version: 0.55
  date: 2005-04-14
  changes:
    lib/emitter.c: 
    - output inline collections.  ensure proper quoting and no blocks in inline collections.
    - safe indentation of scalar blocks which include YAML document separators.
    
    lib/syck.h: styles for every type of node now.

    lib/gram.y: restructuring, trying to fix problem with typed seq-in-map shortcut nodes.
    
    lib/token.re: fixed parsing of inline seq closing char as closing map char and vice versa.
    
    ext/ruby/lib/yaml/rubytypes.rb: added styling of collections.  to_yaml_style for
      every object in Ruby.
    
    ext/ruby/ext/syck/rubyext.c: ditto.

- version: 0.54
  date: 2005-04-07
  changes:
    lib/emitter.c: scale back double-quoting of strings.  get empty nulls working.

    lib/gram.y: allowing transfers, anchors on an empty.

    lib/implicit.re: removing 'y' and 'n' again!!

    lib/yaml.rb: repair YAML.transfer to work with taguris.

    lib/yaml/tag.rb:
    - allow taguri to be set using an accessor.
    - continue support of Object#to_yaml_type.

    ext/ruby/ext/syck/rubyext.c:
    - fixing PrivateType and DomainType objects.
    - new Resolver#tagurize converts type ids to taguris.
      (for backwards compatibility, since the new stuff uses strictly taguri.)
    - merging nobu's changes from Ruby trunk.

    ext/ruby/lib/yaml/rubytypes.rb: simplify taguri display.

- version: 0.53
  date: 2005-03-28
  changes:
    README.EXT: more docs.

    ext/ruby/ext/syck/rubyext.c:
    - moved Object#yaml_new into the node_import and made it the
      default behavior.
    - the target_class is always called wih yaml_new, prepended
      a parameter, which is the klass.
    - loaded nodes through GenericResolver show their style.
    - YAML::Syck::Node#transform works.
    - bug with YAML::Syck::Emitter#level

    ext/ruby/lib/yaml.rb: reworking YAML::Stream to use the new
      emitter.

    ext/ruby/lib/yaml/stream.rb: ditto.

    ext/ruby/lib/yaml/rubytypes.rb: use of new yaml_new syntax.

    ext/ruby/lib/yaml/tag.rb: the tag_subclasses? method now
      shows up in the class.

    ext/ruby/yts/yts.rb: moved eval code out of the YAML
      module.
    ext/ruby/yts/*.yml: prepend YAML module onto all
      code that uses Stream, Omap, etc.

    lib/implicit.re: were 'y' and 'n' seriously omitted??

    lib/node.c: added syck_seq_assign.

- version: 0.52
  date: 2005-03-27
  changes:
    README.EXT: added detailed API docs for Syck!!

    ext/ruby/ext/syck/rubyext.c: consolidated all the diaspora of internal 
      node types into the family below YAML::Syck::Node -- Map, 
      Seq, Scalar -- all of whom are SyckNode structs pointing to 
      Ruby data.
    
    lib/emitter.c: consolidated redundant block_styles struct into
      the scalar_style struct.  (this means loaded nodes can now
      be sent back to emitter and preserve at least its very basic
      formatting.)

    lib/token.re: ditto.

    lib/node.c: new syck_replace_str methods and syck_empty_*
      methods for rewriting node contents, while keeping the ID
      and other setup info.

    lib/syck.h: reflect block_styles and new node functions.

- version: 0.51
  date: 2005-02-09
  changes:
    tests/YTS.c: tests failing due to bad syck_emit_scalar call.

    ext/ruby/lib/yaml/rubytypes.rb: added Object#yaml_new.

    ext/ruby/ext/syck/rubyext.c: both yaml_new and yaml_initialize get
      called, should they be present.
    
- version: 0.50
  date: 2005-02-08
  changes:
    lib/gram.y: freed up parser to accept any structure as a headless
      document!

    lib/emitter.c: renovated emitter, which allows an initial walk
      of the tree to check for anchors and tag prefixing.  then, swift
      emission of nodes with folding routines and shortcut-checking
      built-in!

   ext/ruby/ext/syck/rubyext.c: added hooks for the new emitter.

- version: 0.45
  date: 2004-08-18
  changes:
    CHANGELOG: Added.

    lib/Makefile.am: re2c compiling with bit vectors now.

    lib/syck.c: clear parser on init. thanks, ts. [ruby-core:02931]

    lib/implicit.re: 
    - added sexagecimal float#base60.

    lib/token.re: 
    - using newline_len to handline CR-LFs.  
      "\000" was showing up on folded blocks which
      stopped at EOF. 

    - buffer underflow. thanks, ts. [ruby-core:02929]

    - indentation absolutely ignored when processing flow collections.

    - plain scalars are trimmed if indentation follows in an ambiguous
      flow collection.

    - fixed problem with comments sharing line with opening of block sequence.

    ext/ruby/lib/yaml/compat.rb: moved all ruby 1.6 -> ruby 1.8
      handling into compat.rb.

    ext/ruby/lib/yaml/baseemitter.rb: 
    - simpler flow block code.

    - was forcing a mod value of zero at times, which kept 
      some blocks from getting indentation.

    - double-quoted strings now are handled through the flow block code to
      increase readability.

    ext/ruby/lib/yaml/rubytypes.rb: 
    - exceptions were using an older YAML.object_maker. [ruby-core:03080]
    - subtleties in emitting strings with opening whitespace.
       
    ext/ruby/lib/yaml.rb: 

ext/lua/syck.c
/*
 * syck.c
 *
 * $Author$
 * $Date$
 *
 * Copyright (C) 2005 Zachary P. Landau <kapheine@divineinvasion.net>
 * slact touched this in 2008. blame him.
 */

#include <syck.h>
#include <string.h>
#include <stdlib.h>

#include "lauxlib.h"
#include "lua.h"
#include "lualib.h"

struct emitter_xtra {
	lua_State *L;
	luaL_Buffer output;
	int id;
};

struct parser_xtra {
	lua_State *L; //syck parser thread state
	lua_State *orig; //original API stack state
};

/*
static void stackDump (lua_State *L) {
      int i;
      int top = lua_gettop(L);
      for (i = 1; i <= top; i++) {  // repeat for each level 
        int t = lua_type(L, i);
		printf("%i:	", (top-i+1));
		switch (t) {
    
          case LUA_TSTRING:  // strings 
            printf("[string]");
            break;
    
          case LUA_TBOOLEAN:  // booleans 
            printf(lua_toboolean(L, i) ? "true" : "false");
            break;
    
          case LUA_TNUMBER:  // numbers 
            printf("%g", lua_tonumber(L, i));
            break;
    
          default:  // other values 
            printf("%s", lua_typename(L, t));
            break;
    
        }
        printf("\n");  // put a separator 
      }
      printf("\n----------\n");  // end the listing 
	  }
*/
SYMID
lua_syck_parser_handler(SyckParser *p, SyckNode *n)
{
	struct parser_xtra *bonus = (struct parser_xtra *)p->bonus;
	int o, o2, o3 = -1;
	SYMID oid;
	int i;

	if(!lua_checkstack(bonus->L, 1))
		luaL_error(bonus->orig,"syck parser wanted too much stack space.");
	
	switch (n->kind) {
		case syck_str_kind:
			if (n->type_id == NULL || strcmp(n->type_id, "str") == 0) {
				lua_pushlstring(bonus->L, n->data.str->ptr, n->data.str->len);
				o = lua_gettop(bonus->L);
			}
			else if (strcmp(n->type_id, "null") == 0)
			{
				lua_pushnil(bonus->L);
				o = lua_gettop(bonus->L);
			}
			else if (strcmp(n->type_id, "bool#yes") == 0)
			{
				lua_pushboolean(bonus->L, 1);
				o = lua_gettop(bonus->L);
			}
			else if (strcmp(n->type_id, "bool#no") == 0)
			{
				lua_pushboolean(bonus->L, 0);
				o = lua_gettop(bonus->L);
			}
			else if (strcmp(n->type_id, "float") == 0 || strcmp(n->type_id, "float#fix") == 0 || strcmp(n->type_id, "float#exp") == 0) 
			{
				double f;
				syck_str_blow_away_commas(n);
				f = strtod(n->data.str->ptr, NULL);
				lua_pushnumber(bonus->L, f);
				o = lua_gettop(bonus->L);
			}
			else if (strcmp(n->type_id, "int#hex") == 0)
			{
				long intVal = strtol(n->data.str->ptr, NULL, 16);
				lua_pushnumber(bonus->L, intVal);
				o = lua_gettop(bonus->L);
			}
			else if (strcmp(n->type_id, "int") == 0)
			{
				long intVal = strtol(n->data.str->ptr, NULL, 10);
				lua_pushnumber(bonus->L, intVal);
				o = lua_gettop(bonus->L);                 
			}
			else
			{
				lua_pushlstring(bonus->L, n->data.str->ptr, n->data.str->len);
				o = lua_gettop(bonus->L);
			}
			break;

		case syck_seq_kind:
			lua_newtable(bonus->L);
			o = lua_gettop(bonus->L);
			for ( i=0; i < n->data.list->idx; i++ )
			{
				oid = syck_seq_read(n, i);
				syck_lookup_sym(p, oid, (char **)&o2);
				lua_pushvalue(bonus->L, o2);
				lua_rawseti(bonus->L, o, i+1);
			}
			break;

		case syck_map_kind:
			lua_newtable(bonus->L);
			o = lua_gettop(bonus->L);
			for ( i=0; i < n->data.pairs->idx; i++ )
			{
				oid = syck_map_read(n, map_key, i);
				syck_lookup_sym(p, oid, (char **)&o2);
				oid = syck_map_read(n, map_value, i);
				syck_lookup_sym(p, oid, (char **)&o3);

				lua_pushvalue(bonus->L, o2);
				lua_pushvalue(bonus->L, o3);
				lua_settable(bonus->L, o);
			}
			break;
	}
	oid = syck_add_sym(p, (char *)o);
	return oid;
}

void lua_syck_emitter_handler(SyckEmitter *e, st_data_t data)
{
	struct emitter_xtra *bonus = (struct emitter_xtra *)e->bonus;
	int type = lua_type(bonus->L, -1);
	char buf[32];		/* find a better way, if possible */
	
	switch (type)
	{
		case LUA_TBOOLEAN:
			if (lua_toboolean(bonus->L, -1))
				strcpy(buf, "true");
			else
				strcpy(buf, "false");
			syck_emit_scalar(e, "boolean", scalar_none, 0, 0, 0, (char *)buf, strlen(buf));
			break;
		case LUA_TSTRING:
			syck_emit_scalar(e, "string", scalar_none, 0, 0, 0, (char *)lua_tostring(bonus->L, -1), lua_strlen(bonus->L, -1));
			break;
		case LUA_TNUMBER:
			; lua_Number lnum; //is it an int or a float?
			lnum = lua_tonumber(bonus->L, -1);
			int asInt = lnum;
			if(asInt == lnum)
				snprintf(buf, sizeof(buf), "%i", asInt);
			else
			{
				snprintf(buf, sizeof(buf), "%f", lnum);
				/* Remove trailing zeroes after the decimal point */
				int k;
				for (k = strlen(buf) - 1; buf[k] == '0' && buf[k - 1] != '.'; k--)
					buf[k] = '\0';
			}
			syck_emit_scalar(e, "number", scalar_none, 0, 0, 0, buf, strlen(buf));
			break;
		case LUA_TTABLE:
			if (luaL_getn(bonus->L, -1) > 0) {			/* treat it as an array */
				syck_emit_seq(e, "table", seq_none);
				lua_pushnil(bonus->L);  /* first key */
				while (lua_next(bonus->L, -2) != 0) {
					/* `key' is at index -2 and `value' at index -1 */
					syck_emit_item(e, bonus->id++);
					lua_pop(bonus->L, 1);  /* removes `value'; keeps `key' for next iteration */

				}
				syck_emit_end(e);
			} else {									/* treat it as a map */
				syck_emit_map(e, "table", map_none);
				lua_pushnil(bonus->L);
				while (lua_next(bonus->L, -2) != 0) {
					lua_pushvalue(bonus->L, -2);
					syck_emit_item(e, bonus->id++);
					lua_pop(bonus->L, 1);
					syck_emit_item(e, bonus->id++);
					lua_pop(bonus->L, 1);
				}
				syck_emit_end(e);
			}
			break;
	}

	bonus->id++;
}

static void lua_syck_mark_emitter(SyckEmitter *e, int idx)
{
	struct emitter_xtra *bonus = (struct emitter_xtra *)e->bonus;
	int type = lua_type(bonus->L, idx);

	switch (type) {
		case LUA_TTABLE:
			lua_pushnil(bonus->L);  /* first key */
			while (lua_next(bonus->L, -2) != 0) {
				/* `key' is at index -2 and `value' at index -1 */
				//syck_emitter_mark_node(e, bonus->id++);
				syck_emitter_mark_node(e, bonus->id++);
				lua_syck_mark_emitter(e, -1);
				lua_pop(bonus->L, 1);
			}
			break;
		default:
			syck_emitter_mark_node(e, bonus->id++);
			break;
	}
}


void lua_syck_output_handler(SyckEmitter *e, char *str, long len)
{
	struct emitter_xtra *bonus = (struct emitter_xtra *)e->bonus;
	luaL_addlstring(&bonus->output, str, len);
}

void lua_syck_error_handler(SyckParser *p, char *msg)
{
	//struct parser_xtra *bonus = (struct parser_xtra *)p->bonus;
	luaL_error(((struct parser_xtra *)p->bonus)->orig, "Error at [Line %d, Col %d]: %s\n", p->linect, p->cursor - p->lineptr, msg );
}


static int syck_load(lua_State *L)
{
	struct parser_xtra *bonus;
	SyckParser *parser;
	SYMID v;
	int obj;

	if (!luaL_checkstring(L, 1))
		luaL_typerror(L, 1, "string");

	parser = syck_new_parser();
	parser->bonus = S_ALLOC_N(struct parser_xtra, 1);

	bonus = (struct parser_xtra *)parser->bonus;
	bonus->orig = L;
	bonus->L = lua_newthread(L);

	syck_parser_str(parser, (char *)lua_tostring(L, 1), lua_strlen(L, 1), NULL);
	syck_parser_handler(parser, lua_syck_parser_handler);
	syck_parser_error_handler(parser, lua_syck_error_handler);
	v = syck_parse(parser);
	syck_lookup_sym(parser, v, (char **)&obj);

	syck_free_parser(parser);

	lua_pop(L,1); //pop the thread, we don't need it anymore.
	lua_xmove(bonus->L, L, 1);

	if ( parser->bonus != NULL )
		S_FREE( parser->bonus );
	
	return 1;
}



static int syck_dump(lua_State *L)
{
	SyckEmitter *emitter;
	struct emitter_xtra *bonus;

	emitter = syck_new_emitter();
	emitter->bonus = S_ALLOC_N(struct emitter_xtra, 1);


	bonus = (struct emitter_xtra *)emitter->bonus;
	bonus->L = lua_newthread(L);
	luaL_buffinit(L, &bonus->output);

	syck_emitter_handler(emitter, lua_syck_emitter_handler);
	syck_output_handler(emitter, lua_syck_output_handler);

	lua_pushvalue(L, -2);
	lua_xmove(L, bonus->L, 1);

	bonus->id = 1;
	lua_syck_mark_emitter(emitter, bonus->id);

	bonus->id = 1;
	syck_emit(emitter, bonus->id);
	syck_emitter_flush(emitter, 0);

	luaL_pushresult(&bonus->output);
	
	syck_free_emitter(emitter);
	
	if (bonus != NULL )
		S_FREE( bonus );
	

	return 1;
}

static const luaL_reg sycklib[] = {
	{"load",	syck_load },
	{"dump",	syck_dump },
	{NULL, NULL}
};


static void set_info (lua_State *L)
{
// Assumes the table is on top of the stack.
	lua_pushliteral (L, "_COPYRIGHT");
	lua_pushliteral (L, "Copyright (C) why the lucky stiff");
	lua_settable (L, -3);
	lua_pushliteral (L, "_DESCRIPTION");
	lua_pushliteral (L, "YAML handling through the syck library");
	lua_settable (L, -3);
	lua_pushliteral (L, "_VERSION");
	lua_pushliteral (L, "0.56");
	lua_settable (L, -3);
}

LUALIB_API int luaopen_syck(lua_State *L)
{
	luaL_openlib(L, "syck", sycklib, 0);
	set_info(L);
	return 1;
}